plugins/obs-outputs/ftl-stream.c

```diff
@@ -0,0 +1,1188 @@
+				i * sizeof(*first));
+		if (cur->type == OBS_ENCODER_VIDEO && !cur->keyframe) {
+			*first = *cur;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static void check_to_drop_frames(struct ftl_stream *stream, bool pframes)
+{
+	struct encoder_packet first;
+	int64_t buffer_duration_usec;
+	size_t num_packets = num_buffered_packets(stream);
+	const char *name = pframes ? "p-frames" : "b-frames";
+	int priority = pframes ?
+		OBS_NAL_PRIORITY_HIGHEST : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = pframes ?
+		stream->pframe_drop_threshold_usec :
+		stream->drop_threshold_usec;
+
+	if (num_packets < 5) {
+		if (!pframes)
+			stream->congestion = 0.0f;
+		return;
+	}
+
+	if (!find_first_video_packet(stream, &first))
+		return;
+
+	/* if the amount of time stored in the buffered packets waiting to be
+	 * sent is higher than threshold, drop frames */
+	buffer_duration_usec = stream->last_dts_usec - first.dts_usec;
+
+	if (!pframes) {
+		stream->congestion = (float)buffer_duration_usec /
+			(float)drop_threshold;
+	}
+
+	if (buffer_duration_usec > drop_threshold) {
+		debug("buffer_duration_usec: %" PRId64, buffer_duration_usec);
+		drop_frames(stream, name, priority, pframes);
+	}
+}
+
+static bool add_video_packet(struct ftl_stream *stream,
+		struct encoder_packet *packet)
+{
+	check_to_drop_frames(stream, false);
+	check_to_drop_frames(stream, true);
+
+	/* if currently dropping frames, drop packets until it reaches the
+	 * desired priority */
+	if (packet->priority < stream->min_priority) {
+		stream->dropped_frames++;
+		return false;
+	} else {
+		stream->min_priority = 0;
+	}
+
+	stream->last_dts_usec = packet->dts_usec;
+	return add_packet(stream, packet);
+}
+
+
+static void ftl_stream_data(void *data, struct encoder_packet *packet)
+{
+	struct ftl_stream    *stream = data;
+
+	struct encoder_packet new_packet;
+	bool                  added_packet = false;
+
+	if (disconnected(stream) || !active(stream))
+		return;
+
+	if (packet->type == OBS_ENCODER_VIDEO)
+		obs_parse_avc_packet(&new_packet, packet);
+	else
+		obs_encoder_packet_ref(&new_packet, packet);
+
+	pthread_mutex_lock(&stream->packets_mutex);
+
+	if (!disconnected(stream)) {
+		added_packet = (packet->type == OBS_ENCODER_VIDEO) ?
+			add_video_packet(stream, &new_packet) :
+			add_packet(stream, &new_packet);
+	}
+
+	pthread_mutex_unlock(&stream->packets_mutex);
+
+	if (added_packet)
+		os_sem_post(stream->send_sem);
+	else
+		obs_encoder_packet_release(&new_packet);
+}
+
+static void ftl_stream_defaults(obs_data_t *defaults)
+{
+	UNUSED_PARAMETER(defaults);
+}
+
+
+static obs_properties_t *ftl_stream_properties(void *unused)
+{
+	UNUSED_PARAMETER(unused);
+
+	obs_properties_t *props = obs_properties_create();
+	obs_properties_add_int(props, "peak_bitrate_kbps",
+			obs_module_text("FTLStream.PeakBitrate"),
+			1000, 10000, 500);
+
+	return props;
+}
+
+
+static uint64_t ftl_stream_total_bytes_sent(void *data)
+{
+	struct ftl_stream *stream = data;
+
```

Comment:
There is a blank newline here that isn't mirrored in similar functions below.  For consistency, I'd suggest removing this newline, or adding one to each of the below functions.